sw/otbn/crypto/p256.s

```diff
@@ -1691,6 +1693,133 @@ p256_scalar_mult:
 
   ret
 
+/**
+ * Generate a nonzero random value in the scalar field.
+ *
+ * Returns t, a random value in the range [1,n-1].
+ *
+ * This follows the method in FIPS 186-4 sections B.4.2 and B.5.2 for
+ * generation of secret scalar values d and k. The computation is:
+ *   do {
+ *     c = RBG(256); // fetch 256b random value
+ *   } while (c >= n - 1)
+ *   return c + 1;
+ *
+ * This implementation handles the unmasked secret value, but the seed is
+ * pulled from RND so it cannot be re-run with the same seed. This method
+ * should not be used for keymgr-derived seeds! However, it masks the secret
+ * scalar before returning so that it can be safely handled e.g. in scalarmult.
+ *
+ * Flags: Flags have no meaning beyond the scope of this subroutine.
+ *
+ * @param[in]  w31:  all-zero
+ * @param[out] w20:  first share of secret scalar t
+ * @param[out] w21:  second share of secret scalar t
+ *
+ * clobbered registers: x2, x3, x20, w20, w21, w29
+ * clobbered flag groups: FG0
+ */
+p256_random_scalar:
+  /* Load the curve order n.
+     w29 <= dmem[p256_n] = n */
+  li        x2, 29
+  la        x3, p256_n
+  bn.lid    x2, 0(x3)
+
+  /* w22 <= w29 - 1 = n - 1 */
+  bn.subi   w22, w29, 1
+
+  generate_key_retry:
+  /* Obtain 256 bits of randomness from RND. */
+  bn.wsrr   w20, 0x1 /* RND */
+
+  /* Compare the random value to (n-1).
+     FG0.C <= w20 < w21 = w20 < n - 1 */
+  bn.cmp    w20, w21
+
+  /* Read the FG0.C flag.
+     x2 <= FG0.C = w21 < n - 1 */
+  csrrw     x2, 0x7c0, x0
+  andi      x2, x2, 1
+
+  /* Done if w20 < n - 1, otherwise retry */
+  li        x3, 1
+  bne       x2, x3, generate_key_retry
+
+  /* If we get here, then w20 < n - 1. Add 1 to get the private key.
+     w20 <= w20 + 1 = t */
+  bn.addi   w20, w20, 1
+
+  /* MOD <= n */
+  bn.wsrw   0, w29
+
+  /* Get a new 256-bit random number from URND for masking.
+      w21 <= URND() mod n = t1 */
+  bn.wsrr   w21, 2
```

Comment:
nit: `0x2 /* URND */`  for consistency with other `wsrr` operations. It would be nice if we could pull the auto generated constants like with do in ibex assembly. Not sure if it is possible with the otbn toolchain.

Reply:
Auto-generated constants were discussed as a possible feature for the assembler at some point but aren't yet supported. For now I'll just add the comment.